File: shared/src/main/scala/org/pac4j/lagom/scaladsl/transport/Unauthorized.scala
package org.pac4j.lagom.scaladsl.transport

import com.lightbend.lagom.scaladsl.api.transport.{ExceptionMessage, TransportErrorCode, TransportException}

/**
  * Exception thrown when a service call is unauthorized.
  *
  * @author Sergey Morgunov
  * @since 2.2.0
  */
final class Unauthorized(errorCode: TransportErrorCode, exceptionMessage: ExceptionMessage, cause: Throwable)
    extends TransportException(errorCode, exceptionMessage, cause) {
  def this(errorCode: TransportErrorCode, exceptionMessage: ExceptionMessage) = this(errorCode, exceptionMessage, null)
}

object Unauthorized {
  val ErrorCode: TransportErrorCode = TransportErrorCode(401, 4401, "Unauthorized")

  def apply(message: String) = new Unauthorized(
    ErrorCode,
    new ExceptionMessage(classOf[Unauthorized].getSimpleName, message),
    null
  )

  def apply(exceptionMessage: ExceptionMessage) = new Unauthorized(
    ErrorCode,
    exceptionMessage,
    null
  )

  def apply(cause: Throwable) = new Unauthorized(
    ErrorCode,
    new ExceptionMessage(classOf[Unauthorized].getSimpleName, cause.getMessage),
    cause
  )
}

File: shared/src/main/java/org/pac4j/lagom/javadsl/transport/Unauthorized.java
package org.pac4j.lagom.javadsl.transport;

import com.lightbend.lagom.javadsl.api.deser.ExceptionMessage;
import com.lightbend.lagom.javadsl.api.transport.TransportErrorCode;
import com.lightbend.lagom.javadsl.api.transport.TransportException;

/**
 * Exception thrown when a service call is unauthorized.
 *
 * @author Sergey Morgunov
 * @since 2.2.0
 */
public class Unauthorized extends TransportException {

    private static final long serialVersionUID = 1L;

    public static final TransportErrorCode ERROR_CODE = TransportErrorCode.fromHttp(401);

    public Unauthorized(String message) {
        super(ERROR_CODE, message);
    }

    public Unauthorized(Throwable cause) {
        super(ERROR_CODE, cause);
    }

    public Unauthorized(TransportErrorCode errorCode, ExceptionMessage exceptionMessage) {
        super(errorCode, exceptionMessage);
    }
}
